server/game/core/ability/abilityTargets/TargetResolverBaseClass.ts

```diff
@@ -0,0 +1,55 @@
+import { ITargetResolverBase } from '../../../TargetInterfaces';
+import { AbilityContext } from '../AbilityContext';
+import * as Contract from '../../utils/Contract';
+import { GameSystem } from '../../gameSystem/GameSystem';
+import PlayerOrCardAbility from '../PlayerOrCardAbility';
+import { RelativePlayer } from '../../Constants';
+
+/**
+ * Base class for all target resolvers.
+ */
+export abstract class TargetResolverBaseClass<TProps extends ITargetResolverBase<AbilityContext>> {
+    protected dependentTarget = null;
+    protected dependentCost = null;
+    public constructor(protected name: string, protected properties: TProps, ability: PlayerOrCardAbility) {
+        if (this.properties.dependsOn) {
+            const dependsOnTarget = ability.targetResolvers.find((target) => target.name === this.properties.dependsOn);
+
+            // assert that the target we depend on actually exists
+            Contract.assertNotNullLike(dependsOnTarget);
+
+            dependsOnTarget.dependentTarget = this;
```

Comment:
Hmm, this might be a bug in the engine - since a target can have multiple dependent targets this probably ought to be an array. That would involve a number of other changes though, so I would put a guard here that doesn't use `Contract` but maybe instead just uses an `if` statement that throws an error saying "Multiple dependent targets not implemented yet."

I'm _pretty_ sure no card we have implemented yet would hit that error, but if so it would be good to know now and we can figure out where to go from there.

Reply:
Why should we not use a Contract? I put one in becuase I'm not sure how to throw an error otherwise and nothing tripped it.

Reply:
In my mind they mean slightly different things. Contract failure means "you have put the game in an illegal state and have a bug to fix." Whereas this would be saying "there is a missing feature we need that needs to be added, and you just happened to run into it."

For a while I had a similar kind of thing in TriggeredAbilityWindow because we didn't have the full logic for resolving multiple triggers. In this case it would look like:

```typescript
if (dependsOnTarget.dependentTarget != null) {
    throw new Error("Having multiple dependent targets for the same source target has not yet been implemented");
}

dependsOnTarget.dependentTarget = this;
```

Reply:
To elaborate on why the library is called "Contract" - in my work we refer to those kinds of checks as "contract checks," meaning that it is checking that the software contracts - e.g. between, two objects that are passing data back and forth - have not been violated.